File: 211. Design Add and Search Words Data Structure/211.cpp
/*
Trie Data Strucuture + recursive helper function
m = word length
time complexity:
- addWord: O(m)
- search: 
    - O(m) -> searching defined words
    - O(N * 26^m ) -> worst case situation of searching undefined word with all .
space complexity:
- addWord: O(m)
- search: 
    - O(1) -> searching well defined words
    - O(m) -> recursion stack
*/
class TrieNode{
public:
    bool endofword;
    unordered_map<char, TrieNode*> children;

    TrieNode(bool isend){
        this->endofword = isend;
    }
};

class WordDictionary {
public:
    TrieNode* root;

    WordDictionary() {
        root = new TrieNode(false);
    }
    
    void addWord(string word) {
        TrieNode* currNode = root;
        for(char ch: word){
            if (currNode->children.find(ch) == currNode->children.end()){
                // not found
                currNode->children[ch] = new TrieNode(false);
            }

            currNode = currNode->children[ch];
        }

        currNode->endofword = true;
    }
    
    bool search(string word) {
        return searchintree(word, this->root);
    }

    bool searchintree(string word, TrieNode* currNode){
        // base case is when word length = 1 but repeats the same operation as below
        int i = 0;
        while( i < word.length()){
            char curchar = word[i];
            if (curchar != '.'){
                if (currNode->children.find(curchar) == currNode->children.end()){
                    return false;
                }
                currNode = currNode->children[curchar];
            } else{
                for (auto it: currNode->children){
                    char key = it.first;
                    TrieNode* node = it.second;
                    if (searchintree(word.substr(i+1), node)){
                        return true;
                    }
                }
                return false;
            }
            i++;
        }

        // final check if it is end of word
        return currNode->endofword;
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

File: 211. Design Add and Search Words Data Structure/211.py
"""
Trie: search words of pattern 
time: 
    n = len(word)
    addWord: O(n)
    search: O(26^n)
space: 
    addWord: O(n)
    search: O(n) recursive stack
"""
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class WordDictionary:

    def __init__(self):
        self.rootNode = TrieNode()

    def addWord(self, word: str) -> None:
        curNode = self.rootNode
        for char in word:
            if char not in curNode.children:
                childNode = TrieNode()
                curNode.children[char] = childNode
            else:
                childNode = curNode.children[char]
            
            curNode = childNode
        
        curNode.isEnd = True

    def search(self, word: str) -> bool:
        def findMatchingWord(curIndex, curNode):
            curchar = word[curIndex]
            if curchar != '.':
                if curchar in curNode.children:
                    childNode = curNode.children[curchar]
                    if curIndex == len(word) - 1:
                        if childNode.isEnd == True:
                            return True
                        else:
                            return False
                    else:
                        return findMatchingWord(curIndex + 1, childNode)
                else:
                    return False

            else:
                for char, childNode in curNode.children.items():
                    if curIndex == len(word) - 1:
                        if childNode.isEnd == True:
                            return True
                    else:
                        if findMatchingWord(curIndex + 1, childNode):
                            return True

                return False 

        return findMatchingWord(0, self.rootNode)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

"""
class TrieNode:
    children = {nextchar: TrieNode}
    isEnd = True if there is a word that ends in this TrieNode and otherwise false

init:
    rootNode: TrieNode(isEnd = True)
addWord:
    curNode: rootNode
    for char in word:
        if char not in curNode's children:
            create a child TrieNode
            update curNode.children[char] = childNode
        else:
            access the childNode

        curNode = childNode

    curNode.isEnd = True

search:
    return findMatchingWord(0, rootNode)
    def findMatchingWord(curIndex, curNode):
        if word[curIndex] != '.':
            if word[curIndex] in curNode,children:
                if curIndex == len(word) - 1:
                    if childNode.isEnd == True:
                        return True
                    else:
                        return False

                else:
                    if findMatchingWord(curIndex + 1, curNode.childNode):
                        return True
            else:
                return False
        else:
            for childNode in curNode.children:
                if curIndex == len(word) - 1:
                    if childNode.isEnd == True:
                        return True
                else:
                    if findMatchingWord(curIndex + 1, childNode):
                        return True

            return False
"""
